[#2] Repository

jwonyLee opened the pull request:
## 개요

`PostEntity`를 다루는 데 사용하는 `Repository` 클래스 구현

## 작업 사항

- `PostEntity`와 대응하는 모델 객체 구현
- `create(_:completion:)` 메서드 구현
- `update(_:completion:)` 메서드 구현
- `get(withIdentifier:completion:)` 메서드 구현
- `fetchAll(completion:)` 메서드 구현
- `deleteAll(completion:)` 메서드 구현

`delete` 메서드를 구현하지 못했는데, 이후에 휴지통 기능 구현하면서 같이 구현하겠습니다. 😂

### Linked Issue

close #2

jwonyLee committed:
[#2] test: Post를 하나 만들 수 있는 테스트 작성

jwonyLee committed:
[#2] feat: Post 구조체 생성

jwonyLee committed:
[#2] feat: create new Post

jwonyLee committed:
[#2] feat: Merge branch 'upstream/develop'

jwonyLee committed:
[#2] feat: save Post in CoreData

jwonyLee committed:
[#2] feat: delete All & fetch all Post in CoreData

jwonyLee committed:
[#2] feat: get post with uuid in CoreData

jwonyLee committed:
[#2] feat: update post in Core Data

jwonyLee committed:
[#2] chore: Merge branch from ThingLog/develop

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -5,18 +5,96 @@
+                // then: 예상한 결과 확인
+                switch result {
+                case .success(_):
+                    XCTAssertTrue(true)
```
⚠️ 이 경우는 3개중에 가장 처음에 생성된 경우에만 test성공 여부를 호출 하는 코드같습니다. 3개다 생성한 후에 test를 호출해야할 듯 싶어요.

gustn3965 commented on ThingLog/Model/Attachment.swift:
```diff
@@ -0,0 +1,40 @@
+
+extension Attachment {
+    struct ImageData {
+        let blob: UIImage
```
👀 `blob` 보다는 `originImage` 네이밍이 더 괜찮을 것 같은데 어떠신가욥?!

jwonyLee replied:
`originImage`가 나을까요, `originalImage`가 나을까요?

gustn3965 replied:
`originalImage`로 해욥~

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -0,0 +1,251 @@
+                // then: 예상한 결과 확인
+                switch result {
+                case .success(_):
+                    XCTAssertTrue(true)
```
⚠️ 이 경우는 3개중에 가장 처음에 생성된 경우에만 test성공 여부를 호출 하는 코드같습니다. 3개다 생성한 후에 test를 호출해야할 듯 싶어요.  추가적으로 비동기적으로 실행되는 메서드들이므로 `timeout` 도 호출할 필요가 있습니당

gustn3965 commented on ThingLog/Repository/PostRepository.swift:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    func fetchAll(completion: @escaping (Result<[PostEntity], Error>) -> Void) {
+        coreDataStack.performBackgroundTask { context in
```
👀 `fetchAll`은 언제 사용하나요? 삭제하기 위해 사용하나욥 ?  삭제하기 위해서라면 `bacgroundTask`도 괜찮지만 실제로 데이터를 보여주는 상황에서 read하는 경우이므로 `main Context`에서 호출하는게 더 좋지 않을까 생각합니당

jwonyLee replied:
현재까지는 삭제가 되었는 지 확인하기 위해서만 쓰이고 있어요. 현수님이 `NSFetchedResultController`를 쓰시면 필요 없어지는 메서드이기도 한데 일단 mainContext에서 처리하는 걸로 변경하겠습니당~

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -0,0 +1,251 @@
+                    fatalError(error.localizedDescription)
+                }
+            }
+
```
⚠️ `create`와 `fetchAll`은 둘다 비동기적으로 `background `큐에서 실행하는 걸로 봤습니다. 그러므로 이 중간사이에 그대로 동기적으로 코드를 작성하면, `create`되기 전에 `fetchAll`이 먼저 호출될 수 있습니다. 그러므로 `create`한 다음 해당 클로저 부분안에서 `fetchAll`을 작성해야 한다고 생각합니다.  워낙 데이터가 적기도 하므로 당장은 항상 success가 호출될 수도 있지만 안전하지 않은 코드입니당. 
⚠️ 또한, `테스트 코드 목적이 하나를 만들 수 있다`. 인데, fetchAll에서 result가 나온다고 해서 바로 test를 `true`로 호출하는 것도 잘못된 것 같습니당. `fetchAll`에서는 PostEntity가 없어도 빈 어레이로 반환할 것이기도 하구용.

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -0,0 +1,251 @@
+                    fatalError(error.localizedDescription)
+                }
+            }
+
```
⚠️ 이 부분도 마찬가지로 비동기적으로 코드를 작성해야할 것 같습니다.

gustn3965 commented on ThingLog/Repository/PostRepository.swift:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    func get(withIdentifier identifier: UUID, completion: @escaping (Result<PostEntity, Error>) -> Void) {
+        coreDataStack.performBackgroundTask { context in
```
👀실제로 데이터를 보여주는 상황에서 read하는 경우이므로` main Context`에서 호출하는게 더 좋지 않을까 생각합니당

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -0,0 +1,251 @@
+                }
+            }
+        }
+
```
⚠️ 이 부분도 마찬가지로 비동기적으로 코드를 작성해야할 것 같습니다.

gustn3965 commented on ThingLog/Repository/PostRepository.swift:
```diff
@@ -0,0 +1,104 @@
+    func update(_ updatePost: Post, completion: @escaping (Result<Bool, Error>) -> Void) {
+        coreDataStack.performBackgroundTask { context in
+            do {
+                let request: NSFetchRequest = PostEntity.fetchRequest()
```
👀` identifier로` 하나의 `post`를 찾는 메소드는 `get`메서드와 동일한 로직이라고 생각되는데, `get`을 사용하여 코드를 줄이는 건 어떨까용?  만약 그렇게 한다면, `save`할때 `context`를 맞추어야 해요. get에서도 `performBackgroundTask`는 다른 백그라운드 큐를 이용하므로,  `udpate`내에서도 `performBackgroundTask`가 사용됐으니 서로 다른 context이여서욥

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -0,0 +1,251 @@
+                        switch result {
+                        case .success(_):
+                            // then: 예상한 결과 확인
+                            self.postRepository.get(withIdentifier: newPost.identifier) { result in
```
👍 네 이렇게 클로저안에서 작성을 하거나, perform를 사용하지 않으면서 동기적으로 작성하시면 돼요!

gustn3965 commented on ThingLog/Model/Post.swift:
```diff
@@ -0,0 +1,40 @@
+
+import CoreData
+import Foundation
+
```
⚠️ `Post` 객체 뿐만 아니라 전반적으로 설명이 너무 없습니당 🥲.  저희가 어느정도 추상화한 정보를 알기 때문에 코드도 간결하게 잘 작성하셔서 이해가 가능하지만,  시간이 지나서나, 나중을 위해서 약간의 주석을 추가해주는게 좋다라고 생각합니다.  주석없이 코드가 읽히는게 가장 좋은 코드라고 하긴 하지만, 왜 `Post` 객체가 필요한지에 대한 설명을 추가해주시면 좋겠습니다. 사실 저도 `**Entity`와 `context`가 없는 객체를 따로 만드는 얘기를 듣지 않았다면 왜 Post객체가 따로 필요한지 이해하는데 있어서 어려움이 있었을 것 같아요.   만약 내가 아닌 다른 사람이 처음 이 코드를 본다고 가정했을 때, 주석 없이 한번에 이해가 가능한 코드인지 아닌지를 생각해주시면 좋을 것 같습니당 🙏🏻

gustn3965 commented on ThingLog/Repository/PostRepository.swift:
```diff
@@ -0,0 +1,104 @@
+import UIKit
+
+protocol PostRepositoryProtocol {
+    func create(_ newPost: Post, completion: @escaping (Result<Bool, Error>) -> Void)
```
⚠️`Result` 타입에서 `Error`가 아닌, 아래에 작성한 `PostErpositoryError`로 하는게 좋을 것 같습니다.  지금은 에러 case가 하나밖에 없어서 큰 문제는 되지 않지만, 추후에 추가적인 case가 생길 경우, 그에 맞는 error를 분기처리 해야할텐데, 그때 error를 `타입캐스팅`을 진행해야하기 때문입니다. 마찬가지로, Test코드에 작성한 부분에서도  error발생할 경우 `error.localizedDescription` 을 호출했는데 현재 error를 `타입캐스팅` 하지 않았고, 또한 `PostErpositoryError`에 `LocalizedError `프로토콜을 채택하여 구현하지 않았기 때문에 아무런 정보를 얻을 수 없습니다. error는 발생했지만, 어떤 에러인지를 확인할 수가 없죠.

gustn3965 reviewed (commented):
👀 이 부분은 단순히 궁금하거나, 제안을 드려서 무시하셔도 되는 부분 입니다. 
⚠️ 이 부분은 다시 리팩토링을 진행했으면 하는 부분 입니당. 

`Context`없이 객체를 생성하는 구조는 정말 좋은 구조인 것 같습니다 ! 👍 좋은 구조 배우고 가구용, 코드도 최대한 간결하게 작성하셔서 읽는데 있어서 큰 어려움은 없었네용  열심히 코드 작성하신게 보입니다! 고생하셨어용!! 코멘트들 다시 확인해주세욥~

jwonyLee committed:
[#2] fix: 피드백 반영
- Attachment.ImageData.blob -renaming-> Attachment.ImageData.originalImage
- fetchAll : mainContext.perform 으로 변경
- RepositoryTests 변경
- get : mainContext.perform 으로 변경
- update : get 메서드 사용하는게 변경
- 주석 추가
- Result<_, Error> -> Result<_, PostRepositoryError>로 변경

gustn3965 commented on ThingLog/Repository/PostRepository.swift:
```diff
@@ -0,0 +1,129 @@
+    case failedFetch
+    case failedCreate
+    case failedDelete
+}
```
👀Error를 `PostRepositoryError`로 바꾸셔서 이제 편하게 캐스팅을 할 수 있지만, `LocalizedError`를 채택하고 구현은 하지 않으셨어용 ㅎㅎ.. ` errorDescription` 프로퍼티를 구현해서 각 case마다 error 정보를 반환하는 것도 좋을 것 같습니다. 지금에야 크게 필요가 없지만, 테스트 하는 코드에서 `localizedDescription`이 유용하거든용

jwonyLee replied:
앗 프로토콜 채택하고서 딱히 무조건 구현하라는 부분이 없어서 구현해야 하는 줄 몰랐네요...! 수정하겠습니당

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -0,0 +1,261 @@
+
+        timeout(10) { exp in
+            // when: 테스트중인 코드 실행
+            newPosts.forEach { post in
```
⚠️ 👀오잉 이부분은 수정을 안하신 것 같아용.  forEach가 동기적으로 수행하긴 하지만... 여기까지!

jwonyLee replied:
`forEach`로 create 하면서 성공할 때마다 `postCount`를 증가시키고, `postCount`로 fetch count 비교를 하고 있는데도 추가적인 수정이 필요한가요?_?...

gustn3965 replied:
`forEach`는 동기적으로 수행하지만 `create`는 비동기적으로 이루어지지 않나엽?

gustn3965 commented on ThingLogTests/ThingLogRepositoryTests.swift:
```diff
@@ -0,0 +1,261 @@
+                        switch result {
+                        case .success(let postEntity):
+                            XCTAssertEqual(postEntity.title ?? "", newPost.title)
+                            exp.fulfill()
```
👀 `exp.fulfill()` 이거는 비동기 호출이 끝난 다음에 호출하는게 좋을 것 같아요. 만약에 failure를 호출해서 실패하는 코드를 작성하고 싶은 경우에는 `exp.fulfill()` 이 호출되지 않으니 어떻게 될지 잘 모르겠네용

gustn3965 commented on ThingLog/SupportingFiles/ThingLog.xcdatamodeld/ThingLog.xcdatamodel/contents:
```diff
@@ -17,32 +17,27 @@
         <relationship name="comments" optional="YES" toMany="YES" deletionRule="Cascade" destinationEntity="Comment" inverseName="post" inverseEntity="Comment"/>
-        <relationship name="like" optional="YES" maxCount="1" deletionRule="Cascade" destinationEntity="Like" inverseName="post" inverseEntity="Like"/>
         <relationship name="rating" maxCount="1" deletionRule="Cascade" destinationEntity="Rating" inverseName="post" inverseEntity="Rating"/>
         <relationship name="type" maxCount="1" deletionRule="Cascade" destinationEntity="PostType" inverseName="post" inverseEntity="PostType"/>
```
👀` relationShip` 이름을  `type`  보다는 `postTyp`으로 하는게 어떨까용?  `update` 하는 코드에서도 `.type.type `이렇게 동일한 이름으로 접근하기도 하구욥

gustn3965 reviewed (commented):
고생하셨습니당 ~  
수정해주신 주석 너무 좋은 것 같아용! 
다만 `Post` 객체 뿐만 아니라 다른 `Model`에 대해서도 추상화한 내용을 주석으로 살짝 언급해주시면 더 좋을 것 같습니당  🙏🏻 ㅎㅎ

gustn3965 commented:
앗, 그리구 추가적으로 `Category` 들만 `fetch` 할 수 있도록 메서드를 만들어주시면 감사하겠습니당.  🙏🏻  모아보기 - 카테고리 - 카테고리 선택했을 때 필요합니당

jwonyLee commented:
> 앗, 그리구 추가적으로 `Category` 들만 `fetch` 할 수 있도록 메서드를 만들어주시면 감사하겠습니당. 🙏🏻 모아보기 - 카테고리 - 카테고리 선택했을 때 필요합니당
